wrpclient/wrpclient.go

```diff
@@ -0,0 +1,111 @@
+/**
+ * Copyright 2022 Comcast Cable Communications Management, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package wrpclient
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/xmidt-org/httpaux/erraux"
+	"github.com/xmidt-org/wrp-go/v3"
+)
+
+var (
+	errEncoding             = errors.New("encoding error")
+	errCreateRequest        = errors.New("http request creation error")
+	errHTTPTransaction      = errors.New("http transaction error")
+	errDecoding             = errors.New("decoding response errror")
+	errNonSucessfulResponse = errors.New("non-200 response")
+	errInvalidRequestFormat = errors.New("invalid client RequestFormat")
+)
+
+type HTTPClient interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+type Client struct {
+
+	// URL is the full location for the serverside wrp endpoint.
+	// If unset, use talaria's URI at localhost, which is the port used in talaria's docker image
+	URL string
+
+	// RequestFormat is the wrp Format to use for all requests, which specifies the wrp.Encoder.
+	// If unset, defaults to JSON, which is what the wrp package defaults to.
+	RequestFormat wrp.Format
+
+	// If unset, defaults to net/http.DefaultClient
+	HTTPClient HTTPClient
+}
+
+func (c *Client) checkClientConfig() error {
+	if c.URL == "" {
+		c.URL = "http://localhost:6200"
+	}
+	if c.HTTPClient == nil {
+		c.HTTPClient = &http.Client{}
+	}
+	if c.RequestFormat == 0 {
+		c.RequestFormat = 1
+	} else if c.RequestFormat > 2 || c.RequestFormat < 0 {
+		return errInvalidRequestFormat
```

Comment:
Personally, I would rather have a `NewClient()` function for failures that are consistent and related to the client's configuration.  Then an application can fail on startup rather than continuously failing at runtime when every `SendWRP()` call errors out.  The URL stays constant, so we could also add more validation of that to ensure that it is at least parseable.  @johnabass, what do you think?

Reply:
Agreed, a `NewClient(...) (*Client, error)` is more in keeping with modern `golang`, as opposed to when the stdlib was written.

There are a few reasons to prefer this constructor approach, where you construct a `Client` once and reuse it:

- Errors can short-circuit startup.  @kristinapathak already articulated that well so I won't belabor it :)

- Resources used by the `wrp` client can be reused across multiple requests.  This doesn't really apply right now, but it soon will given some other features we want to add.

- It particpates in dependency injection easily.  Consider two approaches, both using uber/fx:
```
// there is no NewClient:
app := fx.New(
    fx.Provide(
        func() *Client {
            return &Client{ /* stuff * }
        },
    ),
)

// now consider if there is a NewClient function
// just the removal of a couple of lines of code makes a big difference in readability
app := fx.New(
    fx.Provide(
        wrp.NewClient,
    ),
)
```

- The Client's state can be better encapsulated, preventing changes to state that are allowed by the language but are really anti-patterns.  Consider a race condition that can happen trivially:
```
client := &Client{ /* stuff */ }
go client.SendWRP(...)
client.HTTPClient = UhOhRaceCondition // this can't ever happen if HTTPClient is unexported and immutable
```